### PR TITLE
Mainnet upgrade test

### PR DESCRIPTION
**Pull request by faze-quadrata**
## Checklist
1. [ ] Unit Testing?
2. [ ] Backward Compatible?
3. [ ] Upgrade Safe (storage)?
4. [ ] Updated Confluence?
5. [ ] Updated Gitbook?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -10,4 +10,3 @@ types
 #Hardhat files
 cache
 artifacts
-.openzeppelin
```
[the remark this answers is not in the document]
Comment (faze-quadrata) [reply .gitignore:13]: have to copy paste mainnet.json over to unknown-31337.json to work each run

File: test/misc/test_passport_migration.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+const {
+    GOVERNANCE_ROLE,
+    ATTRIBUTE_DID,
+    ATTRIBUTE_AML,
+    ATTRIBUTE_COUNTRY,
+    ATTRIBUTE_IS_BUSINESS,
+} = require("../../utils/constant.ts");
+
+describe.skip("PassportMigration", async () => {
```
Comment (fabrice-quadrata) [inline test/misc/test_passport_migration.ts:16]: can we rename this and the filename to passport_upgrade

File: test/misc/test_passport_migration.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    describe("upgrade", async () => {
+        it("succeed", async () => {
```
Comment (fabrice-quadrata) [inline test/misc/test_passport_migration.ts:18]: can we split the tests into three distinct categories:

1) 5 previous individual passports + 5 business passports can still be queried with the exact same values
2) a preapproved wallet can still query  before and after upgrade
3) a non-preapproved wallet CANNOT query before and after upgrade
4) all QuadGovernance storage values remain the same  before and after upgrade
5) all QuadPassport storage values remain the same  before and after upgrade
6) an admin can still update some QuadGovernance values  before and after upgrade
7) a non-admin CANNOT update some admin QuadGovernance values before and after upgrade
8) an operator can still call setTokenURI after upgrade
9) all assigned roles for DEFAULT_ADMIN, GOVERNANCE, READER, OPERATOR, ISSUER remain the same before and after upgrade
10) an unassigned role remain unassigned for DEFAULT_ADMIN, GOVERNANCE, READER, OPERATOR, ISSUER remain the same before and after upgrade
11) QuadPassport.attributes & QuadPassport.attribute cannot be called by a NON-READER role before and after upgrade
12) QuadGOvernance.upgrade, QuadREader.upgrade, QuadPasport.upgrade cannot be called by non-governance role

File: test/misc/test_passport_migration.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                    },
+                ],
+            });
+            // Preapproved address
+            await network.provider.request({
+              method: "hardhat_impersonateAccount",
+              params: ["0xbA80003B975D1C82d64Efc7F8d2daC393E79B3f9"],
+            });
+
+            const quadPassport = await ethers.getContractAt("QuadPassport", "0x2e779749c40CC4Ba1cAB4c57eF84d90755CC017d");
```
Comment (fabrice-quadrata) [inline test/misc/test_passport_migration.ts:37]: read all values from mainnet/data

File: test/misc/test_passport_migration.ts (hunk 2 above)
Comment (faze-quadrata) [reply test/misc/test_passport_migration.ts:16]: ok

File: test/misc/test_passport_migration.ts (hunk 3 above)
Comment (faze-quadrata) [reply test/misc/test_passport_migration.ts]: done

File: test/misc/test_passport_upgrade.ts [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+const assertGovernanceValues = async (quadGovernance) => {
+    expect(await quadGovernance.treasury()).eql('0xa011eB50e03CaeCb9b551Df9Df478b6a513e0d21')
+    expect((await quadGovernance.revSplitIssuer()).toString()).eql('50')
+    expect(await quadGovernance.passport()).eql('0x2e779749c40CC4Ba1cAB4c57eF84d90755CC017d')
+    expect((await quadGovernance.getIssuersLength()).toString()).eql('3')
+    expect((await quadGovernance.getAllIssuersLength()).toString()).eql('3')
+    expect((await quadGovernance.getEligibleAttributesLength()).toString()).eql('4')
+    expect(await quadGovernance.getIssuers()).eql(['0x38a08d73153F32DBB2f867338d0BD6E3746E3391','0xA095585b1EF2310B4EcBe198a6A6CB86Ef386aBF','0x7256a9eE71fFFc02a92CAbBf950ea6e27f71bBF5'])
+    expect(await quadGovernance.issuersTreasury('0x38a08d73153F32DBB2f867338d0BD6E3746E3391')).eql('0x5F3f69808772C56Daee7A5d3176990733C67A123')
+    expect(await quadGovernance.issuersTreasury('0xA095585b1EF2310B4EcBe198a6A6CB86Ef386aBF')).eql('0xb93b22B75ac3EA6B5066c169B747DF249034F467')
```
Comment (fabrice-quadrata) [inline test/misc/test_passport_upgrade.ts:101]: Can we assert that previously pre approved address are still approved?

Can we assert that previously non-pre-approved adddsss are still non-approved?

File: test/misc/test_passport_upgrade.ts [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+            await quadGovernance.connect(timelock).grantRole(OPERATOR_ROLE, operator.address)
+
+            // Non-admin cannot set role
+            await expect(quadGovernance.connect(preapproved).grantRole(GOVERNANCE_ROLE, deployer.address)).to.be.revertedWith('AccessControl: account 0xba80003b975d1c82d64efc7f8d2dac393e79b3f9 is missing role 0x0000000000000000000000000000000000000000000000000000000000000000')
+
+            // Not preapproved addresses should fail pre-upgrade
+            await expect(quadReader.connect(timelock).callStatic.getAttributes(INDIVIDUAL_ADDRESS_1, ATTRIBUTE_DID)).to.be.revertedWith('SENDER_NOT_AUTHORIZED')
+
+            // Assert roles pre-upgrade
+            expect(await quadGovernance.hasRole(GOVERNANCE_ROLE, timelock.address)).eql(true)
```
Comment (fabrice-quadrata) [inline test/misc/test_passport_upgrade.ts:181]: Can we actually create a separate function for roles and assert before and after upgrade that nothing changes?

What I suggest is to have a list of all addresses and have a mapping of which adddsss should have which one and also assert that addresses shouldn't have the other roles.

File: test/misc/test_passport_upgrade.ts [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+            // Assert non-readers cannot call QuadPassport.attributes()
+            await expect(upgradedPassport.connect(deployer).attributes(deployer.address, ATTRIBUTE_AML)).to.be.revertedWith('INVALID_READER')
+
+            await assertGovernanceValues(upgradedGovernance)
+            await assertPassportValues(upgradedPassport)
+
+            // Should not be able to call upgrade from non-governance
+            await expect(upgradedPassport.connect(preapproved).upgradeTo(deployer.address)).to.be.revertedWith('INVALID_ADMIN')
+            await expect(upgradedGovernance.connect(preapproved).upgradeTo(deployer.address)).to.be.revertedWith('INVALID_ADMIN')
+            await expect(upgradedReader.connect(preapproved).upgradeTo(deployer.address)).to.be.revertedWith('INVALID_ADMIN')
```
Comment (fabrice-quadrata) [inline test/misc/test_passport_upgrade.ts:280]: Can we add ALL admin functions to ensure that it reverts if not admin

File: test/misc/test_passport_upgrade.ts [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+            expect(await upgradedGovernance.hasRole(DEFAULT_ADMIN_ROLE, timelock.address)).eql(true)
+            expect(await upgradedGovernance.hasRole(READER_ROLE, readerOnlyAddr)).eql(true)
+
+            // Assert burn address does not have any roles post-upgrade
+            expect(await upgradedGovernance.hasRole(GOVERNANCE_ROLE, '0x0000000000000000000000000000000000000000')).eql(false)
+            expect(await upgradedGovernance.hasRole(DEFAULT_ADMIN_ROLE, '0x0000000000000000000000000000000000000000')).eql(false)
+            expect(await upgradedGovernance.hasRole(READER_ROLE, '0x0000000000000000000000000000000000000000')).eql(false)
+
+            // Assert non-readers cannot call QuadPassport.attributes()
+            await expect(upgradedPassport.connect(deployer).attributes(deployer.address, ATTRIBUTE_AML)).to.be.revertedWith('INVALID_READER')
```
Comment (fabrice-quadrata) [inline test/misc/test_passport_upgrade.ts:272]: There's also the normal "attribute"

File: test/misc/test_passport_upgrade.ts [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+            // Admin can still call grantRole after upgrade
+            await upgradedGovernance.connect(timelock).grantRole(GOVERNANCE_ROLE, deployer.address)
+            // Non-admin cannot set role
+            await expect(upgradedGovernance.connect(preapproved).grantRole(GOVERNANCE_ROLE, deployer.address)).to.be.revertedWith('AccessControl: account 0xba80003b975d1c82d64efc7f8d2dac393e79b3f9 is missing role 0x0000000000000000000000000000000000000000000000000000000000000000')
+
+            // Assert wallets have same DID, AML and COUNTRY post-upgrade
+            for (const wallet of [INDIVIDUAL_ADDRESS_1, INDIVIDUAL_ADDRESS_2, INDIVIDUAL_ADDRESS_3, INDIVIDUAL_ADDRESS_4, INDIVIDUAL_ADDRESS_5, BUSINESS_ADDRESS_1]){
+                let results = await fetchResults(upgradedReader, preapproved, wallet)
+                expect(results['did'][0].value).eql(EXPECTED_ADDRESS_RESULTS[wallet]['did'])
```
Comment (fabrice-quadrata) [inline test/misc/test_passport_upgrade.ts:253]: Can we also add a test to make sure that a non-passport holder remains a non-passport holder

File: test/misc/test_passport_upgrade.ts [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+            expect(await upgradedGovernance.hasRole(READER_ROLE, '0x0000000000000000000000000000000000000000')).eql(false)
+
+            // Assert non-readers cannot call QuadPassport.attributes()
+            await expect(upgradedPassport.connect(deployer).attributes(deployer.address, ATTRIBUTE_AML)).to.be.revertedWith('INVALID_READER')
+
+            await assertGovernanceValues(upgradedGovernance)
+            await assertPassportValues(upgradedPassport)
+
+            // Should not be able to call upgrade from non-governance
+            await expect(upgradedPassport.connect(preapproved).upgradeTo(deployer.address)).to.be.revertedWith('INVALID_ADMIN')
```
Comment (fabrice-quadrata) [inline test/misc/test_passport_upgrade.ts:278]: Can we add test for all private method to make sure that they are still private (aka cannot call) for both QuadPassport and QuadReader and QuadGovernance

File: test/misc/test_passport_upgrade.ts [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+    expect(await quadGovernance.getIssuers()).eql(['0x38a08d73153F32DBB2f867338d0BD6E3746E3391','0xA095585b1EF2310B4EcBe198a6A6CB86Ef386aBF','0x7256a9eE71fFFc02a92CAbBf950ea6e27f71bBF5'])
+    expect(await quadGovernance.issuersTreasury('0x38a08d73153F32DBB2f867338d0BD6E3746E3391')).eql('0x5F3f69808772C56Daee7A5d3176990733C67A123')
+    expect(await quadGovernance.issuersTreasury('0xA095585b1EF2310B4EcBe198a6A6CB86Ef386aBF')).eql('0xb93b22B75ac3EA6B5066c169B747DF249034F467')
+    expect(await quadGovernance.issuersTreasury('0x7256a9eE71fFFc02a92CAbBf950ea6e27f71bBF5')).eql('0xa011eB50e03CaeCb9b551Df9Df478b6a513e0d21')
+}
+
+/// To get this test to work, you have to copy/paste .openzeppelin/mainnet.json into unknown-31337.json
+/// or just run in the project root dir:
+///
+/// cp -r .openzeppelin/mainnet.json .openzeppelin/unknown-31337.json; npx hardhat test test/misc/test_passport_upgrade.ts
```
Comment (faze-quadrata) [inline test/misc/test_passport_upgrade.ts:108]: @fabrice-quadrata

File: test/misc/test_passport_upgrade.ts (hunk 6 above)
Comment (fabrice-quadrata) [reply test/misc/test_passport_upgrade.ts]: see `scripts/verify/assertMainnetDeployment.ts`

File: test/misc/test_passport_upgrade.ts (hunk 9 above)
Comment (faze-quadrata) [reply test/misc/test_passport_upgrade.ts]: https://quadrata.atlassian.net/browse/QUAD-1450

File: test/misc/test_passport_upgrade.ts (hunk 10 above)
Comment (faze-quadrata) [reply test/misc/test_passport_upgrade.ts]: https://quadrata.atlassian.net/browse/QUAD-1450

**Review by fabrice-quadrata — APPROVED** [review]

**Pull request merged**